Microsoft.PowerShell.Crescendo/src/experimental/HelpParsers/README.md

```diff
@@ -0,0 +1,249 @@
+
+- the executable name and how to retrieve help (is it `-?`, `--help`, or something else)
+- the pattern which designates that sub-commands will follow
+- the pattern which designates that options will follow
+- the pattern which designates that arguments will follow
+- the pattern which designates that usage will follow
+- the pattern which designates the actual parameter
+- the pattern which designates additional help (on-line or other commands)
+
+The help text (for the tools I've chosen so far) is regular enough that when one of these strings is found,
+we start hunting for following data to use.
+For example, when the `Sub-Command` pattern is used, we start a loop which looks for new commands.
+If we find a new command, we call the help parser and start the scan again for the new "command"
+
+Each pattern represents elements which can be incorporated into the configuration.
+
+#### Getting the Help Text
+
+The scanner first starts with the help.
+This means that the scanner has to know two things:
+
+- the name of the executable
+- the parameter to use to get help
+
+There was not much variation here, it was either `-?` and `--help` so each one of the parsers has something similar to:
+
+```powershell
+$exe = "docker"
+$helpChar = "--help"
+```
+
+In order to start the parsing, I need to gather the first level help, so `& $exe $helpChar` is executed as part of the `parseHelp` function.
+
+> **_It's a little more complicated than that_** - 
+The function itself is recursive because I have to parse the help multiple times.
+I'll search for sub-commands and when found, I'll call the `parseHelp` function and add the new sub-command,
+which will parse the _new_ command before the original help is finished parsing.
+Eventually, we'll run out of sub-commands so the help will finally get parsed,
+but if you notice the _order_ in which the parsed help objects are returned, they may not be the order you expect.
+
+#### Parameter Recognition
+
+Sometimes, the command argument would have an associated type for the parameter and sometime that would not be available.
+This is why there are 2 patterns for a parameter:
+
+```powershell
+$parmPattern = "--(?<pname>[-\w]+)\s(?<ptype>\w+)\s+(?<phelp>.*)|--(?<pname>[-\w]+)\s+(?<phelp>.*)"
+```
+
+This pattern also ignores those patterns which are a single character.
+I did this on purpose as PowerShell prefers parameters which are more descriptive.
+I didn't create aliases out of those one character parameters, which is something that could be done in the future.
+I also don't handle the cases where the one character parameter _follows_ the more verbose parameter.
+In my inspection of the help, they all pretty much did something like:
+
+```bash
+	-q, --quiet     Do not be overly verbose
+```
+
+rather than:
+
+```bash
+	--quiet, -q     Do not be overly verbose
+```
+
+So my parameter recognizer expects it and doesn't capture the first part of the string.
+Again, this is probably pretty easy to support.
```

Comment:
Expects what? I am not clear what "it" is referring to here. Also, I tested your pattern and it works the same for either case.